File: kiro-sdk/src/acp-transport.ts
/**
 * ACP Transport — manages the kiro-cli acp child process and JSON-RPC 2.0 communication.
 *
 * Singleton: one long-lived process shared across all sessions.
 */

import { spawn, type ChildProcess } from 'child_process';
import type { AcpInitializeResult } from './types.js';

type PendingRequest = {
  resolve: (value: unknown) => void;
  reject: (error: Error) => void;
  timeout: ReturnType<typeof setTimeout>;
};

type NotificationHandler = (method: string, params: Record<string, unknown>) => void;

export class AcpTransport {
  private process: ChildProcess | null = null;
  private ready = false;
  private rpcId = 0;
  private pending = new Map<number, PendingRequest>();
  private lineBuffer = '';
  private onNotification: NotificationHandler = () => {};
  private initPromise: Promise<AcpInitializeResult> | null = null;

  constructor(private kiroPath = process.env.KIRO_PATH || 'kiro-cli') {}

  /** Register a handler for incoming JSON-RPC notifications. */
  setNotificationHandler(handler: NotificationHandler): void {
    this.onNotification = handler;
  }

  private initResult: AcpInitializeResult | null = null;

  /** Ensure the ACP process is running and initialized. */
  async connect(acpArgs: string[] = []): Promise<AcpInitializeResult> {
    // If process died, reset so we respawn
    if (this.initResult && (!this.ready || !this.process)) {
      this.initResult = null;
      this.initPromise = null;
    }

    if (this.ready && this.process && this.initResult) {
      return this.initResult;
    }

    if (this.initPromise) return this.initPromise;

    this.initPromise = this._spawn(acpArgs);
    return this.initPromise;
  }

  private async _spawn(acpArgs: string[]): Promise<AcpInitializeResult> {
    const args = ['acp', ...acpArgs];

    this.process = spawn(this.kiroPath, args, {
      stdio: ['pipe', 'pipe', 'pipe'],
      env: { ...process.env },
    });

    this.lineBuffer = '';

    this.process.stdout!.on('data', (data: Buffer) => {
      this.lineBuffer += data.toString();
      const lines = this.lineBuffer.split('\n');
      this.lineBuffer = lines.pop()!;
      for (const line of lines) this.handleLine(line);
    });

    this.process.stderr!.on('data', (data: Buffer) => {
      const msg = data.toString().trim();
      if (msg && !msg.includes('DeprecationWarning')) {
        console.error('[kiro-sdk:stderr]', msg);
      }
    });

    this.process.on('close', () => {
      this.ready = false;
      this.process = null;
      this.initPromise = null;
      this.initResult = null;
      for (const [id, { reject, timeout }] of this.pending) {
        clearTimeout(timeout);
        reject(new Error('ACP process exited'));
      }
      this.pending.clear();
    });

    this.process.on('error', (err) => {
      console.error('[kiro-sdk] process error:', err.message);
      this.ready = false;
      this.process = null;
      this.initPromise = null;
      this.initResult = null;
    });

    this.ready = true;

    const result = await this.sendRpc('initialize', {
      protocolVersion: 1,
      clientCapabilities: { fs: { readTextFile: true, writeTextFile: true }, terminal: true },
      clientInfo: { name: 'kiro-sdk', version: '0.1.0' },
    }) as AcpInitializeResult;

    this.initResult = result;
    return result;
  }

  /** Send a JSON-RPC request and wait for the response. */
  sendRpc(method: string, params: Record<string, unknown> = {}): Promise<unknown> {
    return new Promise((resolve, reject) => {
      if (!this.process || !this.ready) {
        return reject(new Error('ACP process not ready'));
      }
      const id = ++this.rpcId;
      const timeout = setTimeout(() => {
        this.pending.delete(id);
        reject(new Error(`RPC timeout for ${method}`));
      }, 120_000);
      this.pending.set(id, { resolve, reject, timeout });
      this.process.stdin!.write(JSON.stringify({ jsonrpc: '2.0', id, method, params }) + '\n');
    });
  }

  /** Gracefully shut down the ACP process. */
  disconnect(): void {
    if (this.process) {
      this.process.kill('SIGTERM');
      this.process = null;
      this.ready = false;
    }
  }

  private handleLine(line: string): void {
    if (!line.trim()) return;
    let msg: Record<string, unknown>;
    try { msg = JSON.parse(line); } catch { return; }

    // Response to a pending request
    if (msg.id != null && this.pending.has(msg.id as number)) {
      const { resolve, reject, timeout } = this.pending.get(msg.id as number)!;
      clearTimeout(timeout);
      this.pending.delete(msg.id as number);
      if (msg.error) {
        const err = msg.error as Record<string, unknown>;
        reject(new Error((err.message as string) || JSON.stringify(err)));
      } else {
        resolve(msg.result);
      }
      return;
    }

    // Notification
    if (msg.method) {
      this.onNotification(msg.method as string, (msg.params || {}) as Record<string, unknown>);
    }
  }
}

File: kiro-sdk-python/kiro_sdk/acp_transport.py
"""ACP Transport — manages the kiro-cli acp subprocess and JSON-RPC 2.0 over stdio."""

from __future__ import annotations

import asyncio
import json
import os
import sys
from typing import Any, Callable

from .types import AcpInitializeResult

NotificationHandler = Callable[[str, dict[str, Any]], None]


class AcpTransport:
    def __init__(self, kiro_path: str | None = None) -> None:
        self._kiro_path = kiro_path or os.environ.get("KIRO_PATH", "kiro-cli")
        self._process: asyncio.subprocess.Process | None = None
        self._ready = False
        self._rpc_id = 0
        self._pending: dict[int, asyncio.Future[Any]] = {}
        self._on_notification: NotificationHandler = lambda m, p: None
        self._init_result: AcpInitializeResult | None = None
        self._reader_task: asyncio.Task[None] | None = None

    def set_notification_handler(self, handler: NotificationHandler) -> None:
        self._on_notification = handler

    async def connect(self, acp_args: list[str] | None = None) -> AcpInitializeResult:
        if self._ready and self._process and self._init_result:
            return self._init_result
        return await self._spawn(acp_args or [])

    async def _spawn(self, acp_args: list[str]) -> AcpInitializeResult:
        args = [self._kiro_path, "acp", *acp_args]
        self._process = await asyncio.create_subprocess_exec(
            *args,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            limit=10 * 1024 * 1024,  # 10 MB — kiro-cli sends large JSON-RPC lines
        )
        self._ready = True
        self._reader_task = asyncio.get_event_loop().create_task(self._read_loop())
        asyncio.get_event_loop().create_task(self._stderr_loop())

        raw = await self.send_rpc("initialize", {
            "protocolVersion": 1,
            "clientCapabilities": {"fs": {"readTextFile": True, "writeTextFile": True}, "terminal": True},
            "clientInfo": {"name": "kiro-sdk-python", "version": "0.1.0"},
        })
        self._init_result = AcpInitializeResult(
            protocol_version=raw.get("protocolVersion", 0),
            agent_capabilities=raw.get("agentCapabilities", {}),
            agent_info=raw.get("agentInfo", {}),
        )
        return self._init_result

    async def send_rpc(self, method: str, params: dict[str, Any] | None = None) -> Any:
        if not self._process or not self._ready:
            raise RuntimeError("ACP process not ready")
        self._rpc_id += 1
        rid = self._rpc_id
        future: asyncio.Future[Any] = asyncio.get_event_loop().create_future()
        self._pending[rid] = future
        payload = json.dumps({"jsonrpc": "2.0", "id": rid, "method": method, "params": params or {}}) + "\n"
        self._process.stdin.write(payload.encode())  # type: ignore[union-attr]
        await self._process.stdin.drain()  # type: ignore[union-attr]
        try:
            return await asyncio.wait_for(future, timeout=120)
        except asyncio.TimeoutError:
            self._pending.pop(rid, None)
            raise TimeoutError(f"RPC timeout for {method}")

    def disconnect(self) -> None:
        if self._process:
            self._process.terminate()
            self._process = None
            self._ready = False
            self._init_result = None
        if self._reader_task:
            self._reader_task.cancel()
            self._reader_task = None

    async def _read_loop(self) -> None:
        assert self._process and self._process.stdout
        try:
            while True:
                line = await self._process.stdout.readline()
                if not line:
                    break
                self._handle_line(line.decode())
        except asyncio.CancelledError:
            pass
        finally:
            self._ready = False
            for fut in self._pending.values():
                if not fut.done():
                    fut.set_exception(RuntimeError("ACP process exited"))
            self._pending.clear()

    async def _stderr_loop(self) -> None:
        assert self._process and self._process.stderr
        try:
            async for line in self._process.stderr:
                msg = line.decode().strip()
                if msg and "DeprecationWarning" not in msg:
                    print(f"[kiro-sdk:stderr] {msg}", file=sys.stderr)
        except asyncio.CancelledError:
            pass

    def _handle_line(self, line: str) -> None:
        line = line.strip()
        if not line:
            return
        try:
            msg = json.loads(line)
        except json.JSONDecodeError:
            return

        rid = msg.get("id")
        if rid is not None and rid in self._pending:
            fut = self._pending.pop(rid)
            if msg.get("error"):
                fut.set_exception(RuntimeError(msg["error"].get("message", str(msg["error"]))))
            else:
                fut.set_result(msg.get("result"))
            return

        if "method" in msg:
            self._on_notification(msg["method"], msg.get("params", {}))
